doc/design/concurrent_data_transfer_and_kernel_execution.md

```diff
@@ -0,0 +1,97 @@
+
+To support the above description, we need to define a new class: `Channel`. Here, we borrow the concept of [`Channel`](https://www.golang-book.com/books/intro/10#section2) in the go language.
+
+```
+template <typename T>
+class Channel {
+ private:
+  using ChannelElement = std::vector<T>;
+
+  std::size_t capacity_;
+  std::size_t bytes_limit_;
+  std::size_t current_bytes_;
+  std::mutex mu_;
+  std::condition_variable empty_cond_var_;
+  std::condition_variable full_cond_var_;
+  std::deque<ChannelElement> channel_;
+
+ public:
+  explicit Channel(std::size_t capacity, std::size_t bytes_limit);
+
+  void Put(ChannelElement* buffer_element) { ... }
+  void Get(ChannelElement* buffer_element) { ... }
+  size_t Size() { ... }
+  void Clear()  { ... }
+};
+};
+```
+
+**Take the [recognize_digits_mlp](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/tests/book/test_recognize_digits_mlp.py) as an example to show the model definition after using the staging area mechanism.**
+
+```
+...
+concurrent_program = Program()
+with program(concurrent_program):
+    image = fluid.layers.data_layer(...)
+    label = fluid.layers.data_layer(...)
+    places = get_places()
+    channel_list = create_channel_list(name="data_buffer")
+    with parallel.for(places) as for_loop:
+        cur_channel = fluid.layers.create_channel(channel_list, for_loop.i, channel_size=2)  
+        # if channel_list has the specific channel, 
+        # this function will return it directly.
+        fluid.layers.send_to_channel(input=[for_loop.input(image), 
+                                for_loop.input(label)], out=cur_channel)
+
+main_program = Program()
+with program(main_program):
+    places = get_places()
+    channel_list = create_channel_list(name="data_buffer")
+    with parallel.for(places) as for_loop:
+        cur_channel = fluid.layers.get_channel(channel_list, for_loop.i)
+        image, label = fluid.layers.receive_from_channel(input=cur_channel)
+        
+        y_predict = fluid.layers.fc(input=image, size=1, act=None)
+        cost = fluid.layers.square_error_cost(input=y_predict, label=label)
+        avg_cost = fluid.layers.mean(x=cost)
+        ....
+
+for i in range(buffer_size):
+    data = next(train_reader())
+    executor.run(concurrent_program, feed=feeder.feed(data))
+
+for pass_id in range(PASS):
+    for data in train_reader():
+        executor.run(concurrent_program, feed=[...])
+        executor.run(main_program, fetch=[...])
+
+for i in range(buffer_size):
+    executor.run(main_program, fetch=[...])
+
+```
+In Python code, we define two `program`, `concurrent_program` used to send data into `Channel` and `main_program` used to get data from the `Channel` and execute training. If you're familiar with [`Goroutine`](https://www.golang-book.com/books/intro/10#section1) in the go language, you'll find that `main_program` and `concurrent_program` just like two `Goroutine`.
```

Comment:
If we make `main_program ` and `concurrent_program ` like two Goroutine, I think we need to declare the code block will be run as a Goroutine,  and pass the channel as a argument, so for my personal opinion, the Python code would look like:

```python
chan = fluid.channel.make(type=var.LoDTensor)

with fluid.go(concurrent_program, chan):
    image = fluid.layers.data_layer(...)
    label = fluid.layers.data_layer(...)
    places = get_places() 
    with parallel.for(places) as for_loop:
        fluid.channle.send(chan, data=[for_loop.input(image), for_loop.input(label)])

with fluid.go(main_program, chan):
    places = get_places() 
    with parallel.for(places) as for_loop:
        image, label = fluid.channel.recv(chan)
        y_predict = fluid.layers.fc(input=image, size=1, act=None)
        cost = fluid.layers.square_error_cost(input=y_predict, label=label)
        avg_cost = fluid.layers.mean(x=cost) 
```